server/app/controllers/userController.js

```diff
@@ -0,0 +1,100 @@
+import UserModel from '../models/userModel';
+import { validateEmail, validatePassword } from '../services/validators';
+import { tokenForUser } from '../services/passport';
+
+export const createUser = async (req, res, next) => {
+  const { email, password, firstName, lastName } = req.body;
+
+  const errorMessages = {};
+
+  const emailTaken = await UserModel.findOne({ email });
+  if (emailTaken) {
+    errorMessages.email = 'This email is already taken.';
+  }
+
+  if (!validateEmail(email)) {
+    errorMessages.email = 'Please enter a valid email.';
+  }
+
+  if (!validatePassword(password)) {
+    errorMessages.password = 'Your password must be at least 6 characters.';
+  }
+
+  const errorType =
+    Object.keys(errorMessages).length > 0 ? 'INVALID_FORM' : null;
+
+  if (errorType !== null) {
+    res.status(400).json({
+      errorType,
+      errorMessages,
+    });
+    return;
+  }
+
+  const newUser = new UserModel({
+    email,
+    password,
+    firstName,
+    lastName,
+  });
+
+  try {
+    const savedUserModel = await newUser.save();
+    res.status(200).json({
+      token: tokenForUser(savedUserModel.id),
+      user: savedUserModel,
+    });
+  } catch (e) {
+    console.error(e);
```

Comment:
Would the idea here be to formally log this error?

Obviously logging is not something we care about now... but just thinking long term and how we wanna do this "right"

Reply:
Sure, we can do really do whatever we want with this - no idea what we want long term.

Reply:
re: logging, a coworker should be this blog today written by founder of heroku, thought you mind find interesting: https://12factor.net/logs

If you haven't seen https://12factor.net/ it's pretty nice at a high description of best practices/scale (but of course nothing is dogma)